### PR TITLE
Add support for retransmit option to a put job for failing blobs

### PR DESCRIPTION
**Pull request by shabtaisharon**
* Need to add more integration tests
* Need to finish go proxy

### CODE REVIEW

File: IntegrationTestDs3WithProxy/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Comment (RachelTucker) [inline IntegrationTestDs3WithProxy/Properties/AssemblyInfo.cs]: missing copyright

File: TestDs3/Helpers/TestDs3ClientHelpers.cs [hunk 2]
```diff
@@ -585,5 +586,404 @@ public void TestJobOnFailureEvent()
...
+            node2Client.VerifyAll();
+            clientFactory.VerifyAll();
+            client.VerifyAll();
+
+            CollectionAssert.AreEquivalent(new[] { 15L, 20L, 11L, 10L, 10L, 10L }, dataTransfers);
+            CollectionAssert.AreEquivalent(Stubs.ObjectNames, itemsCompleted);
+        }
+
+        [Test]
+        public void TestPutJobWithRetransmit2Blobs1Fail()
```
Comment (RachelTucker) [inline TestDs3/Helpers/TestDs3ClientHelpers.cs]: A lot of the setup for the new tests are very similar (if not the same).  Can some of the common functionality be pulled out into a setup function?

File: Ds3/Helpers/Jobs/Job.cs [hunk 3]
```diff
@@ -149,19 +162,25 @@ private void TransferBlob(IDs3Client client, Blob blob)
...
+                    ranges,
+                    stream,
+                    _metadataAccess,
+                    MetadataListener,
+                    _retransmitRetries
+                    );
+            }
+            finally
+            {
+                this._streamFactory.CloseBlob(blob);
```
Comment (rpmoore) [inline Ds3/Helpers/Jobs/Job.cs]: Good catch

File: Ds3/Helpers/Transferrers/PartialDataTransferrerDecorator.cs [hunk 4]
```diff
@@ -15,36 +15,29 @@
 
 using System;
 using System.Collections.Generic;
-using Ds3.Models;
-using Ds3.Helpers.Jobs;
 using System.IO;
+using Ds3.Helpers.Jobs;
```
Comment (rpmoore) [inline Ds3/Helpers/Transferrers/PartialDataTransferrerDecorator.cs]: Is everything in this class just changing the style?

Comment (shabtaisharon) [inline Ds3/Helpers/Transferrers/PartialDataTransferrerDecorator.cs]: looks like it :)

File: Ds3/Helpers/Transferrers/WriteTransferrer.cs [hunk 5]
```diff
@@ -43,7 +37,26 @@ internal class WriteTransferrer : ITransferrer
...
-            client.PutObject(request);
+            try
+            {
+                client.PutObject(request);
+            }
+            catch (Exception ex)
+            {
+                if (retransmitRetries == 0)
+                {
+                    throw new Ds3NoMoreRetransmitException(
```
Comment (rpmoore) [inline Ds3/Helpers/Transferrers/WriteTransferrer.cs]: This should use the same exception that the Get retry code uses.  If need be, update the exception that the Get code throws to match this exception.

File: Ds3/Helpers/Transferrers/WriteTransferrer.cs [hunk 6]
```diff
@@ -43,7 +37,26 @@ internal class WriteTransferrer : ITransferrer
...
+                        string.Format(Resources.NoMoreRetransmitException, request.ObjectName, request.Offset.Value),
+                        ex);
+                }
+
+                if (!stream.CanSeek) throw new Ds3NotSupportedStream(Resources.NotSupportedStream, ex);
+
+                retransmitRetries--;
+                stream.Position = request.Offset.Value;
+                Transfer(client, bucketName, objectName, blobOffset, jobId, ranges, stream, metadataAccess,
```
Comment (rpmoore) [inline Ds3/Helpers/Transferrers/WriteTransferrer.cs]: Are you using recursion to do the retry logic?  If this is the case, this should be updated to use a loop instead.

File: IntegrationTestDs3WithProxy/NonSeekableStream.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            return this._stream.Read(buffer, offset, count);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override bool CanRead { get; }
+        public override bool CanSeek => false;
```
Comment (rpmoore) [inline IntegrationTestDs3WithProxy/NonSeekableStream.cs]: Is this a C# 4.5 feature that we can use?  I thought this was added in C# 6

Comment (shabtaisharon) [inline IntegrationTestDs3WithProxy/NonSeekableStream.cs]: the project uses a 4.5.2 so i think we are good

File: TestDs3/Helpers/TestDs3ClientHelpers.cs (hunk 2 above)
Comment (shabtaisharon) [inline TestDs3/Helpers/TestDs3ClientHelpers.cs]: i've opened a new jira task for this

**Pull request merged**